cli/src/action/database/upgrade/scabbard.rs

```diff
@@ -0,0 +1,231 @@
+    node_id::store::{diesel::DieselNodeIdStore, NodeIdStore},
+    store::ConnectionUri,
+};
+
+/// Migrate all of the service state's current commit hashes to the [`CommitHashStore`].
+pub fn upgrade_scabbard_commit_hash_state(
+    state_dir: &Path,
+    database_uri: &ConnectionUri,
+) -> Result<(), UpgradeError> {
+    let lmdb_db_factory = LmdbDatabaseFactory::new_state_db_factory(state_dir, None);
+    let upgrade_stores = UpgradeStores::new(database_uri)?;
+
+    let node_id = if let Some(node_id) = upgrade_stores
+        .node_id_store
+        .get_node_id()
+        .map_err(|e| InternalError::from_source(Box::new(e)))?
+    {
+        node_id
+    } else {
+        // This node has not even set a node id, so it cannot have any circuits.
+        info!("Skipping scabbard commit hash store upgrade, no local node ID found");
+        return Ok(());
+    };
+
+    let circuits = upgrade_stores
+        .admin_service_store
+        .list_circuits(&[])
+        .map_err(|e| InternalError::from_source(Box::new(e)))?;
+
+    let local_services = circuits
+        .into_iter()
+        .map(|circuit| {
+            circuit
+                .roster()
+                .iter()
+                .filter_map(|svc| {
+                    if svc.node_id() == node_id && svc.service_type() == "scabbard" {
+                        Some((
+                            circuit.circuit_id().to_string(),
+                            svc.service_id().to_string(),
+                        ))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>()
+        })
+        .flatten();
+
+    for (circuit_id, service_id) in local_services {
+        let lmdb_commit_hash_store =
+            TransactCommitHashStore::new(lmdb_db_factory.get_database(&circuit_id, &service_id)?);
+        let db_commit_hash_store = upgrade_stores
+            .commit_hash_store_factory
+            .new_commit_hash_store(&circuit_id, &service_id);
+
+        if let Some(current_commit_hash) = lmdb_commit_hash_store
+            .get_current_commit_hash()
+            .map_err(|e| InternalError::from_source(Box::new(e)))?
+        {
+            db_commit_hash_store
+                .set_current_commit_hash(&current_commit_hash)
+                .map_err(|e| InternalError::from_source(Box::new(e)))?;
+            info!("Upgraded scabbard service {}::{}", circuit_id, service_id);
+        } else {
+            debug!(
+                "No commit hash found for service {}::{}",
+                circuit_id, service_id
+            );
+        }
+    }
+
+    Ok(())
+}
+
+struct UpgradeStores {
+    admin_service_store: Box<dyn AdminServiceStore>,
+    commit_hash_store_factory: CommitHashStoreFactory,
+    node_id_store: Box<dyn NodeIdStore>,
+}
+
+impl UpgradeStores {
+    fn new(database_uri: &ConnectionUri) -> Result<Self, UpgradeError> {
+        match database_uri {
```

Comment:
This match would fail to compile if postgres, sqlite, and memory were all disabled. We should make sure at a minimum that a reasonable compilation error is given.

Reply:
Should we configure it such that the upgrade command is only available if at least sqlite or postgres are enabled?

Reply:
I'm not sure how we would implement that cleanly. 

The default branch - "Unsupported database:" -- that would seem to be a reasonable error from the CLI if we have the databases disabled. But that's only on if memory is on, which I assume is an artifact of it being an enum. I wonder if this enum should be declared non-exhaustive.

Maybe something to handle in a different PR though, or not at all, since this doesn't actually break the build and we are compiling with --no-default-features. You would have to be trying to recreate the feature conditions. Though if we remove node-id-upgrade feature then that may expose it more directly.